### DISCUSSION TITLE
How to execute after dependency when using the add_middleware method in FastAPI?

### DISCUSSION THREAD

**Discussion by NB-123** [fastapi/fastapi#11287] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Code Snippets
from middlewares import record_api_usage_to_db
from fastapi import FastAPI, Request, Depends, status
from starlette.middleware.base import BaseHTTPMiddleware
app = FastAPI()
app.add_middleware(BaseHTTPMiddleware, dispatch= record_api_usage_to_db_middleware)

app.include_router(
admin_router,
tags=["admin"],
prefix="/admin",
dependencies=[Depends(auth_check)],)

async def record_api_usage_to_db_middleware(request: Request, background_tasks: BackgroundTasks, call_next, db: Session = Depends(get_db)):
response = await call_next(request)
background_tasks.add_task(record_api_usage_to_db, db, request)
return response
```

### Description

In my project, I need to protect some API Endpoints. To do this, I have defined an auth check function that checks the authentication and authorization of a particular user.

I then include this as a dependency to all my API routes that need protection like so:

app.include_router(
admin_router,
tags=["admin"],
prefix="/admin",
dependencies=[Depends(auth_check)],)
I now want to track usage stats, and have defined a middleware function with background tasks like so:

async def record_api_usage_to_db_middleware(request: Request, background_tasks: BackgroundTasks, call_next, db: Session = Depends(get_db)):
response = await call_next(request)
background_tasks.add_task(record_api_usage_to_db, db, request)
return response
The above code sits in a separate file called middlewares.py. I then add this to my main.py file where we instantiate the FastAPI object like so:

from middlewares import record_api_usage_to_db
from fastapi import FastAPI, Request, Depends, status
from starlette.middleware.base import BaseHTTPMiddleware
app = FastAPI()
app.add_middleware(BaseHTTPMiddleware, dispatch= record_api_usage_to_db_middleware)
I want this middleware to run only for authenticated endpoints and AFTER the auth has happened since I need some metadata from the authentication/authorization step to be included in the usage data.

I am not sure how to do this, because it seems FastAPI does not allow dependencies in middleware. Any suggestions? Thanks.

### Operating System

macOS

### FastAPI Version

0.110.0

### Pydantic Version

1.10.14

### Python Version

3.10.7

**Comment by YuriiMotov** [comment, marked as answer]
I may be wrong, but it seems like the only way to wrap your endpoints by any code after dependencies are resolved is to use decorator.
```
from functools import wraps
from typing import Awaitable, Callable, ParamSpec, TypeVar
from fastapi import Depends, FastAPI

app: FastAPI = FastAPI()

P = ParamSpec("P")
R = TypeVar("R")

def my_decorator(func: Callable[P, Awaitable[R]]) -> Callable[P, Awaitable[R]]:
    @wraps(func)
    async def wrapper(*args: P.args, **kwargs: P.kwargs) -> R:
        dep_value = kwargs.get("dep_value")
        print(dep_value)
        response = await func(*args, **kwargs)
        return response
    return wrapper

def my_dep() -> int:
    return 123

@app.get("/testing")
@my_decorator
async def testing(dep_value: int = Depends(my_dep)) -> None:
    pass
```

**UPD:** You can also adapt [this example](https://github.com/tiangolo/fastapi/discussions/11294) to your needs. I'm not sure it's a good idea, but it will work for all endpoints:
```
from typing import Any

import fastapi
from fastapi import Depends, FastAPI
from fastapi.dependencies.models import Dependant

async def run_endpoint_function_wrapped(
    *,
    dependant: Dependant,
    values: dict[str, Any],
    is_coroutine: bool,
) -> Any:
    global original_run_endpoint_function
    print(values)
    response = await original_run_endpoint_function(
        dependant=dependant, values=values, is_coroutine=is_coroutine
    )
    return response

app = FastAPI()
original_run_endpoint_function = fastapi.routing.run_endpoint_function
fastapi.routing.run_endpoint_function = run_endpoint_function_wrapped

def my_dep() -> int:
    return 123

@app.get("/")
def index(dep_value: int = Depends(my_dep)):
    pass

```

**Reply by NB-123** [reply]
I see - so what you are saying is that I should not be using middleware to write usage stats into my database, but rather use a decorator. How do I do this with my current syntax?: 
app.include_router(
admin_router,
tags=["admin"],
prefix="/admin",
dependencies=[Depends(auth_check)],)
Or do I need to add the decorator one by one (seems inefficient and defeats the point of the include_router. Thanks

**Reply by YuriiMotov** [reply]
With decorators you have to apply this decorator to every endpoint.
Also consider second variant I added to my answer
